src/grpc_interceptor/testing/dummy_client.py

```diff
@@ -27,55 +28,68 @@ class DummyService(dummy_pb2_grpc.DummyServiceServicer):
             This allows testing special cases, like raising exceptions.
     """
 
-    def __init__(self, special_cases: Dict[str, SpecialCaseFunction]):
+    def __init__(
+        self,
+        special_cases: Dict[str, SpecialCaseFunction],
+        context_cases: Optional[Dict[str, SpecialCaseFunction]] = None,
+    ):
         self._special_cases = special_cases
+        self._context_cases = context_cases
 
     def Execute(
         self, request: DummyRequest, context: grpc.ServicerContext
     ) -> DummyResponse:
         """Echo the input, or take on of the special cases actions."""
-        return DummyResponse(output=self._get_output(request))
+        return DummyResponse(output=self._get_output(request, context))
 
     def ExecuteClientStream(
         self, request_iter: Iterable[DummyRequest], context: grpc.ServicerContext
     ) -> DummyResponse:
         """Iterate over the input and concatenates the strings into the output."""
-        output = "".join(self._get_output(request) for request in request_iter)
+        output = "".join(self._get_output(request, context) for request in request_iter)
         return DummyResponse(output=output)
 
     def ExecuteServerStream(
         self, request: DummyRequest, context: grpc.ServicerContext
     ) -> Iterable[DummyResponse]:
         """Stream one character at a time from the input."""
-        for c in self._get_output(request):
+        for c in self._get_output(request, context):
             yield DummyResponse(output=c)
 
     def ExecuteClientServerStream(
         self, request_iter: Iterable[DummyRequest], context: grpc.ServicerContext
     ) -> Iterable[DummyResponse]:
         """Stream input to output."""
         for request in request_iter:
-            yield DummyResponse(output=self._get_output(request))
+            yield DummyResponse(output=self._get_output(request, context))
 
-    def _get_output(self, request: DummyRequest) -> str:
+    def _get_output(self, request: DummyRequest, context: grpc.ServicerContext) -> str:
         input = request.input
+
+        output = input
         if input in self._special_cases:
             output = self._special_cases[input](input)
-        else:
-            output = input
+        if self._context_cases and input in self._context_cases:
+            output = self._context_cases[input](context)
+
         return output
 
 
 @contextmanager
 def dummy_client(
     special_cases: Dict[str, SpecialCaseFunction],
-    interceptors: List[ServerInterceptor],
+    context_cases: Optional[Dict[str, SpecialCaseFunction]] = None,
```

Comment:
I think we should leave the order as is here, just make `interceptors` optional. The reason is that reordering the parameters is a backwards-incompatible change, so we should only do it if it's a significant improvement.

Reply:
I think this becomes moot with changing the `SpecialCaseFunction` signature to include context? We'd drop `context_cases` anyway which would revert the change param ordering.